src/scanner.c

```diff
@@ -1,205 +1,102 @@
   lexer->advance(lexer, false);
 }
 
-static void skip(TSLexer *lexer) {
-  lexer->advance(lexer, true);
-}
-
-static bool scan_str(TSLexer *lexer) {
-  // We want to delegate the scanning of the start-of-string/end-of-string '"'
-  // character to the grammar defined in grammar.js.
-  // So the idea is we track if we've seen any string content,
-  // and if we see EOF or an unescaped '"' char _and_ we haven't consumed any string content,
-  // we return false to indicate to tree-sitter that our custom scanner has not found
-  // a token.
-  // Additionally, if we come across an escape sequence and we already have content,
-  // we want to mark the end of that content and stop scanning (i.e. return true);
-  // when tree-sitter calls us next, we'll resume from where we left off
-  // (i.e. right at the start of the escape sequence).
-  bool has_content = false;
-  
-  lexer->result_symbol = STR_CONTENT;
-
-  while (true) {
+// Here we only parse literal fragment inside a string.
+// Delimiter, interpolation and escape sequence are handled by the parser and we simply stop at them.
+//
+// The implementation is inspired by tree-sitter-javascript:
+// https://github.com/tree-sitter/tree-sitter-javascript/blob/fdeb68ac8d2bd5a78b943528bb68ceda3aade2eb/src/scanner.c#L19
+static bool scan_string_fragment(TSLexer *lexer) {
+  lexer->result_symbol = STRING_FRAGMENT;
+  for (bool has_content = false;; has_content = true) {
+    lexer->mark_end(lexer);
     switch (lexer->lookahead) {
       case '"':
-        if (has_content) {
-          lexer->mark_end(lexer);
-          return true;
-        } else {
-          return false;
-        }
       case '\\':
-        lexer->mark_end(lexer);
-        if (!has_content) {
-          advance(lexer);
-          // accept anything (other than EOF) following the '\\'
-          if (lexer->eof(lexer)) {
-            return false;
-          }
-          advance(lexer);
-          lexer->result_symbol = ESCAPE_SEQUENCE;
-          lexer->mark_end(lexer);
-        }
-        return true;
+        return has_content;
       case '$':
-        lexer->mark_end(lexer);
         advance(lexer);
         if (lexer->lookahead == '{') {
-          if (has_content) {
-            return true;
-          } else {
-            return false;
-          }
-        } else if (lexer->lookahead != '"' && lexer->lookahead != '\\' ) {
-          // any char following '$' other than '"', '\\' and '{' (which was handled above)
+          return has_content;
+        } else if (lexer->lookahead != '"' && lexer->lookahead != '\\') {
+          // Any char following '$' other than '"', '\\' and '{' (which was handled above)
           // should be consumed as additional string content.
+          // This means `$${` doesn't start an interpolation, but `$$${` does.
           advance(lexer);
-          lexer->mark_end(lexer);
         }
-        has_content = true;
         break;
+      // Simply give up on EOF or '\0'.
```

Comment:
The official Nix parser currently accepts NUL chars, so that might be something we want to preserve (though I'm inclined to accept this PR as-is and consider bringing back the NUL char support later).